File: ios/RideLink/NowPlayingCard.swift
import RideLinkCore
import SwiftUI

/// Play/pause/seek/next/previous — mirrors `com.ridelink.app.ui.NowPlayingCard`.
struct NowPlayingCard: View {
    let playerState: PlayerState
    let currentEntry: LibraryEntry?
    let queueSize: Int
    var title: String? = nil
    var artist: String? = nil
    let onPlay: () -> Void
    let onPause: () -> Void
    let onSeek: (Int64) -> Void
    let onNext: () -> Void
    let onPrevious: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: RideDesign.sm) {
            Text("Now Playing").font(.headline)
            Text(title ?? currentEntry?.track.title ?? (playerState.localEntryId == nil ? "Nothing playing" : "Shared track")).font(.title3.bold())
            if let artist = artist ?? currentEntry?.track.artist, !artist.isEmpty { Text(artist).foregroundStyle(.secondary) }
            if let error = playerState.error { Text(musicFailureLabel(error)).foregroundStyle(.red) }

            if playerState.durationMs > 0 {
                Slider(
                    value: Binding(
                        get: { Double(playerState.positionMs) },
                        set: { onSeek(Int64($0)) }
                    ),
                    in: 0...Double(playerState.durationMs)
                )
                Text("\(formatMs(playerState.positionMs)) / \(formatMs(playerState.durationMs))")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }

            HStack(spacing: RideDesign.md) {
                transport("Previous", icon: "backward.end.fill", action: onPrevious).disabled(queueSize == 0)
                transport(playerState.playing ? "Pause" : "Play", icon: playerState.playing ? "pause.fill" : "play.fill", action: playerState.playing ? onPause : onPlay)
                    .disabled(playerState.localEntryId == nil && currentEntry == nil && !playerState.playing)
                transport("Next", icon: "forward.end.fill", action: onNext).disabled(queueSize == 0)
            }

            Text(queueSize == 0 ? "Queue is empty" : "\(queueSize) items in queue").font(.caption).foregroundStyle(.secondary)
        }
        .padding(RideDesign.lg)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(RideDesign.surface)
        .cornerRadius(RideDesign.radius)
    }

    private func transport(_ label: String, icon: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            VStack(spacing: RideDesign.xs) {
                Image(systemName: icon).font(.title3)
                Text(label).font(.caption)
            }.frame(maxWidth: .infinity, minHeight: RideDesign.touch)
        }.buttonStyle(.bordered).controlSize(.regular).accessibilityLabel(label)
    }

    private func formatMs(_ ms: Int64) -> String {
        let totalSeconds = ms / 1000
        return String(format: "%d:%02d", totalSeconds / 60, totalSeconds % 60)
    }
}

private func musicFailureLabel(_ failure: MusicFailure) -> String {
    switch failure {
    case .decodeFailed: "This audio file could not be played. Try another track."
    case .fileMissing: "The audio file is missing. Import it again."
    case .unsupportedFormat: "This audio format is unsupported."
    case .storageIo: "The audio file could not be read. Check available storage."
    case .cancelled: "Playback cancelled."
    }
}

File: android/app/src/main/kotlin/com/ridelink/app/ui/NowPlayingCard.kt
package com.ridelink.app.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.ridelink.core.library.LibraryEntry
import com.ridelink.core.player.PlayerState

/**
 * Play/pause/seek/next/previous (this phase's brief §21) — REQUIREMENTS §10.2's Ride Mode gets a
 * simplified version of this later (Phase 7); this is the developer/browsing-mode surface, the same
 * scope [MainScreen]'s existing cards are.
 */
@Composable
fun NowPlayingCard(
    playerState: PlayerState,
    currentEntry: LibraryEntry?,
    queueSize: Int,
    onPlay: () -> Unit,
    onPause: () -> Unit,
    onSeek: (Long) -> Unit,
    onNext: () -> Unit,
    onPrevious: () -> Unit,
) {
    Card(modifier = Modifier.fillMaxWidth()) {
        Column(modifier = Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Text("Now Playing", style = MaterialTheme.typography.titleMedium)
            Text(
                currentEntry?.let { "${it.track.title} — ${it.track.artist}" } ?: "Nothing loaded",
                style = MaterialTheme.typography.bodyLarge,
            )
            playerState.error?.let { error ->
                Text("Playback error: ${playerFailureLabel(error)}", style = MaterialTheme.typography.bodySmall)
            }

            if (playerState.durationMs > 0) {
                Slider(
                    value = playerState.positionMs.coerceIn(0, playerState.durationMs).toFloat(),
                    onValueChange = { onSeek(it.toLong()) },
                    valueRange = 0f..playerState.durationMs.toFloat(),
                )
                Text(
                    "${formatMs(playerState.positionMs)} / ${formatMs(playerState.durationMs)}",
                    style = MaterialTheme.typography.labelSmall,
                )
            }

            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                Button(onClick = onPrevious) { Text("Previous") }
                if (playerState.playing) {
                    Button(onClick = onPause) { Text("Pause") }
                } else {
                    Button(onClick = onPlay, enabled = currentEntry != null) { Text("Play") }
                }
                Button(onClick = onNext) { Text("Next") }
            }

            Text("Queue: $queueSize item(s)", style = MaterialTheme.typography.labelSmall)
        }
    }
}

private fun formatMs(ms: Long): String {
    val totalSeconds = ms / MILLIS_PER_SECOND
    val minutes = totalSeconds / SECONDS_PER_MINUTE
    val seconds = totalSeconds % SECONDS_PER_MINUTE
    return "%d:%02d".format(minutes, seconds)
}

private fun playerFailureLabel(failure: com.ridelink.core.player.MusicFailure): String =
    when (failure) {
        com.ridelink.core.player.MusicFailure.DECODE_FAILED -> "could not decode this file"
        com.ridelink.core.player.MusicFailure.FILE_MISSING -> "file not found"
        com.ridelink.core.player.MusicFailure.UNSUPPORTED_FORMAT -> "unsupported format"
        com.ridelink.core.player.MusicFailure.STORAGE_IO -> "storage error"
        com.ridelink.core.player.MusicFailure.CANCELLED -> "cancelled"
        com.ridelink.core.player.MusicFailure.FOREGROUND_SERVICE_START_FAILED -> "bring RideLink to the front and try again"
    }

private const val MILLIS_PER_SECOND = 1000L
private const val SECONDS_PER_MINUTE = 60L
